### DISCUSSION TITLE
네트워크 장치 & 서버 과부하 해결방법

### DISCUSSION THREAD

**Discussion by 24tngus** [24tngus/CS_STUDY#40] [Network]
#### 1. PDU

![](https://velog.velcdn.com/images/24tngus/post/96dec043-d206-440d-966c-0d70d73f6e11/image.png)

#### 2. 계층별 네트워크 장치

- 애플리케이션 계층 : L7 스위치
- 전송 계층 : L4 스위치
- 인터넷 계층(네트워크 계층) : 라우터, L3 스위치
- 데이터링크 계층 : L2 스위치, 브리지
- 물리 계층 : NIC, 리피터, AP

![](https://velog.velcdn.com/images/24tngus/post/ec769894-3e47-4819-a8cd-e5a4a4066279/image.png)

#### 3. 애플리케이션 계층

- 1) 개념
  - 응용 프로세스와 직접 관계하여 서비스 수행
  - 프로토콜 : HTTP, FTP, SMTP, POP3, IMAP, Telnet

- 2) L7 스위치
  - **IP주소, TCP/UDP 포트 정보 및 패킷 내용 **참조하여 스위치
  - L4 스위치 기능 + 컨텐츠 기반 제어
  - 서버 이중화, 보안에 강점 (웹 방화벽 및 보안 스위치에 사용)
  - 서버의 부하를 분산하는 기기 (로드밸런서)
    - IP, Port 뿐만 아니라 url, 헤더, 쿠키 기반으로 트래픽 분산
  - 헬스체크를 통해 장애가 발생한 서버 확인 ➡️ 해당 서버로 트래픽을 보내지 못하게 하는 역할
  ![](https://velog.velcdn.com/images/24tngus/post/7689d40c-8d57-4deb-a279-95694489d8d4/image.png)

> 📖 참고 📖 헬스체크
- 서버의 정상/비정상 여부 판별
- 전송 주기와 재전송 횟수 등을 설정한 후 반복적으로 서버에 요청을 보냄
- L4 또는 L7 스위치에 사용

> 📖 참고 📖 서버 이중화
- 로드 밸런서의 대표적 기능
- 서비스를 안정적으로 이용하기 위해서 2대 이상의 서버 필수적
- 에러가 발생하여 서버 1대 종료해도 안정적으로 운용 목적
![](https://velog.velcdn.com/images/24tngus/post/21077d60-8010-4680-9dde-451e3d6b925c/image.png)

#### 4. 전송 계층

- 1) 개념
  - 두 지점간 신뢰성 있는 데이터 주고받는 역할 (port 번호, 전송방식 결정)
  - 에러 제어, 경로 제어 
  - 프로토콜 : TCP, UDP

- 2) L4 스위치
  - **IP주소, TCP/UDP 포트 정보** 참조하여 스위치
  - L3 스위치 기능 + 로드 밸런싱
  - 서버의 부하를 분산하는 기기 (로드밸런서)
    - 패킷의 IP주소와 Port번호를 참고해서 트래픽 분산
    - TCP, UDP 헤더를 기반으로 우선순위를 판단해서 스위칭 판단 가능
  - 헬스체크를 통해 장애가 발생한 서버 확인
  ![](https://velog.velcdn.com/images/24tngus/post/84b9e843-195c-4bd2-9088-51cf82bb812d/image.png)

#### 5. 네트워크 계층

- 1) 개념
  - 라우터를 통해 경로 선택 + 주소 지정 + 패킷 전달 
  - 단위 : packet
  - 장비 : router

- 2) 라우터
  - 네트워크에서 경로를 선택하는 장비
  - 다른 네트워크에 존재하는 장치끼리 데이터 주고 받을 경우 ➡️ 패킷 소모 최소화, 경로 최소화
  ![](https://velog.velcdn.com/images/24tngus/post/0d59d136-2755-482f-8920-ab15391d5c10/image.png)

- 3) L3 스위치
  - **IP주소** 참조하여 스위치 
  - L2 스위치 기능 + 라우팅
  - 라우팅테이블을 참조해서 IP패킷에 IP주소를 담아 보냄 
  - 패킷 목적지 주소가 라우팅테이블에 없을 경우 패킷 버림

#### 6. 데이터링크 계층

- 1) 개념
  - 송수신되는 정보의 오류와 흐름을 관리하여 통신의 흐름을 관리 (프레임에 MAC주소 부여)
  - 에러 검출, 재전송, 흐름 제어 
  - 단위 : frame
  - 장비 : bridge, switch, Ethernet
  
- 2) L2 스위치
  - **MAC address** 참조하여 스위치
  - 장치들의 MAC 주소를 MAC 주소 테이블을 통해 관리
  - 네트워크 계층에서 받은 패킷을 기반으로 이더넷프레임 만들어 목적지MAC주소로 패킷 보내주는 역할
  ![](https://velog.velcdn.com/images/24tngus/post/eeb006bf-de61-4363-a8ce-d67433cd5b56/image.png)
  
- 3) 브리지
  - 두 개의 LAN을 상호 접속 할 수 있도록 하는 통신망 연결 장치 (포트간 다리 역할)
  - 통신망의 범위 확장 or 서로 다른 LAN을 기반으로 하나의 통신망 구축
  ![](https://velog.velcdn.com/images/24tngus/post/3e98a988-01cc-4077-ac90-df7d6b862b96/image.png)
  
#### 7. 물리 계층

- 1) 개념
  - 무선LAN과 유선LAN을 통해 0/1로 이루어진 데이터를 보내는 계층
  - 단위 : bit
  - 장비 : repeater, hub, 통신 케이블 
  
- 2) NIC (Network Interface Card)
  - LAN 카드
  - 2대 이상의 컴퓨터 네트워크를 구성하는데 사용
  - 네트워크와 빠른 속도로 데이터 송수신 하도록 PC에 설치된 확장 카드
  - 장치에 PC 고유의 식별번호(MAC 주소) 저장
  ![](https://velog.velcdn.com/images/24tngus/post/26a5efe9-76c2-4d3a-a85b-233247dcf415/image.png)

- 3) 리피터
  - 들어오면서 약해진 신호 증폭하여 다른 쪽으로 전달하는 장치
  - 패킷 더 멀리 보냄
  - 광케이블 보급으로 현재 사용X
  
- 4) AP (Access Point)
  - 패킷을 복사하는 장치
  - AP에 유선 LAN을 연결한 후 다른 장치에서 무선 LAN기술(wifi 등)을 사용하여 무선 네트워크 연결 가능
  
> 📖 참고 📖 OSI 7계층 vs TCP/IP 4계층
- OSI 7계층 : 네트워크 전송의 데이터 표준 정림
- TCP/IP 4계층 : 실제 인터넷 통신을 반영한 인테넛 표준
![](https://velog.velcdn.com/images/24tngus/post/6986eee3-a3b1-42ad-962b-ec273c5e4a4b/image.png)

> 📖 참고 📖 유선 LAN
- 동축케이블, 광케이블 기반으로 만들어진 유선LAN을 이루는 이더넷은 **IEEE802.3 프로토콜**을 기반으로 **전이중화 통신**을 사용
- 광역 통신망 기술에도 활용되지만 일반적으로 근거리 통신망 기술
- ✏️ **IEEE802.3**
  - 연결된 이더넷에서 물리 계층, 데이터링크 계층의 매체 접근 제어를 정의하는 Working Group
  - 이더넷프레임을 어떤 구조 기반인지, 케이블의 최대 전송량, 어떤 케이블만 가능한지 등 정한 규칙
  ![](https://velog.velcdn.com/images/24tngus/post/44b4e464-5b9f-42e2-b44a-0bbf79638313/image.png)
- ✏️ **전이중화 통신 (Full Duplex)**
  - 양쪽 장치가 동시에 송수신할 수 있는 방식
  - 송신로, 수신로 나눠서 데이터 주고받음
  - 현대 고속 이더넷의 기반 
- ✏️ **CSMA/CD (Carrier Sense Multiple Access with Collision Detection)**
  - 충돌 감지 기능을 사용해 충돌 여부 확인
  - IEEE 802.3 LAN의 이더넷 전송 프로토콜에서 사용
  - 1️⃣ 회선감지 : 송신 전에 회선이 비어있는지 확인 (Carrier Sense)
  - 2️⃣ 데이터 전송 : 비어있을 경우 네트워크상 모든 컴퓨터가 데이터 전송위해 동시에 액세스 시도 (Multiple Access)
  - 3️⃣ 충돌 발생 : 동시에 데이터 전송할 경우 충돌 발생 후 네트워크 상의 다른 컴퓨터에게 충돌 알림 (Collision Detection)
  - 4️⃣ 회선 감지: 충돌 있을 경우 일정 시간 대기 후 재전송 
  - CSMA방식 단말의 증가에 따라 동시 전송(충돌)할 확률 높아지는 문제점 개선
  ![](https://velog.velcdn.com/images/24tngus/post/0d8fe0a9-ed9e-4d2f-8f21-06d0ab07ca81/image.png)

> 📖 참고 📖 무선 LAN
- IEEE802.11 표준규격을 따르며 **반이중화 통신**을 사용
- 무선랜, wifi 등 무선 근거리 통신망을 위한 컴퓨터 무선 네트워크에 사용되는 기술 
- ✏️ **반이중화 통신 (Half Duplex)**
  - 양쪽 장치는 서로 통신할 수 있지만, 동시에는 통신X
  - 한 번에 한 방향만 통신하는 방식 
  - 장치가 신호를 수신하기 시작하면, 바로 응답할 수 없고 수신이 완료될 때까지 대기 
- ✏️ **CSMA/CA (Carrier Sense Multiple Access with Collision Avoidance)**
  - 충돌 회피 기능 사용
  - 장치에서 데이터 보내기 전 Carrier 감지 등으로 사전에 가능한 충돌 방지 
  - IEEE 802.11 무선 랜 전송 프로토콜에서 사용
  - 1️⃣ 회선 감지 : 송신 전에 회신이 비어있는지 확인 (Carrier Sense)
  - 2️⃣ IFS 대기 :  채널이 idle 해도 IFS 동안 대기하여 전송 늦춰 충돌 피함
    - IFS(InterFrame Space) 프레임간 공간 
    - IFS 작으면 프레임 우선순위 높음
  - 3️⃣ Cotention Window 대기 
    - Contention Window는 time-slot으로 나뉜 일정 시간
    - 전송할 준비가 된 노드는 random 수(0~(2^k-1))만큼 대기 후 프레임 전송
    - 노드는 매 time-slot 뒤에 채널 감지
    - 해당 시간동안 busy 상태면 타이머 중지 후 idle 상태가 되면 다시 타이머 시작 
    - 충돌 가능성 있을 경우 피함 (Collision Avoidcnace)
  - 4️⃣ 데이터 전송
    - 프레임 제대로 받고 + ACK 세그먼트 받을 경우 ➡️ 종료
    - 프레임 받지 못할 경우 ➡️ k = k + 1 위 과정 반복 (k가 정해진 Kmax보다 커지면 해당 프레임 전송 버림)
  ![](https://velog.velcdn.com/images/24tngus/post/48766c3e-ab55-4010-ba7a-df54f5634d31/image.png)

#### 1. 서버 과부하

- 1) 개념
  - 서버가 리소스를 소진하여 들어오는 요청을 처리하지 못함
  - 🗒️ 예시 : 서버는 사용자의 웹요청을 처리하지 못해 응답없음이 뜸
  ![](https://velog.velcdn.com/images/24tngus/post/4c589e73-2159-4742-bc00-c2b002e80611/image.png)
  
- 2) 해결 방법
  - 모니터링을 통한 자원 할당
  - 로드밸런서
  - 블랙스완 프로토콜
  - 서킷 브레이커
  - 컨텐츠 관리

#### 2. 모니터링을 통한 자원 할당

- 1) 원인
  - 자원(CPU, 메모리, 대역폭 등)의 한계점 도달
  - 보통 서버의 CPU 사용량이 80~90%에 도달하거나 메모리가 부족해 계속 스와핑하면 과부하
  
- 2) 모니터링 
  - 서버 과부화로 인한 서버 중지 대처
    - 어떤 페이지에 어떤 트래픽이 얼마나 발생했는지 확인
    - 어떤 네트워크에서 병목현상이 일어났는지 확인
  - 모니터링을 하면 페이지 활용도 높은지/낮은지 파악하여 서비스 개선에 도움
  - 해결하기 위한 문제점 파악
  - [cloudflare](https://www.cloudflarestatus.com/)는 모니터링 결과와 함께 서비스 중단 여부를 사용자에게 알림

- 3) **AWS 오토스케일링**
  - 서비스 이용불가능 상태 발생 이전 cloud watch가 계속 모니터링하여 서버 대수 늘려주는 방법
  - 애플리케이션을 자동으로 모니터링하고 자원의 용량을 자동으로 조정함
  
- 4) **[netdata](https://github.com/netdata/netdata)**를 이용한 모니터링
  - AWS 대체하는 무료 모니터링 서비스 
  - 지속적인 모니터링, 자원할당 해결
  - slack과 연동하여 설정한 임계치를 기반으로 알림서비스 구축 가능
  
#### 3. 로드밸런서

- 1) 개념
  - 서버 앞단에서 로드밸런스 통해 트래픽 분산
  - 한 서버에 장애가 발생하면 트래픽을 다른 기능 서버로 리다이렉션하여 시스템 중단 방지
  ![](https://velog.velcdn.com/images/24tngus/post/e22fca9b-97b2-41c0-b1ba-516529477007/image.png)

  
#### 4. 블랙스완 프로토콜

- 1) 블랙스완 개념
  - 예측할 수 없는 사고가 일어난 것
  - 사후에는 이 사고의 원인 분석할 수 있지만, 사전에는 사고를 예측할 수 없음
  
- 2) 블랙스완 발생시 운영 수칙
  - 영향을 받은 시스템과 각 시스템의 상대적 **위험 수준을 확인**
    - 체계적으로 데이터 수집
    - 원인에 대한 가설을 수립한 후 테스팅
  - 잠재적으로 영향을 받을 수 있는 **내부의 모든 팀에 연락**
  - 최대한 빨리 취약점에 영향을 받는 **모든 시스템을 업데이트**
  - 복원 계획을 포함한 우리의 **대응 과정을** 파트너/고객 등 **외부에 전달** 
  
#### 5. 서킷 브레이커

- 1) 개념
  - 서비스 장애를 감지하고 연쇄적으로 생기는 오류 전파를 방지하는 기법
  - 연쇄적 장애 전파 막기 위해 자동으로 외부 서비스와 연결을 차단/복구
  - 서비스-서비스간 서킷브레이커 계층을 둠
  - 연속적인 에러를 막아 일부 서비스 종료되더라도 다른 서비스 동작 가능
  
- 2) 서킷 브레이커 동작과정
![](https://velog.velcdn.com/images/24tngus/post/072e2f6b-ab36-45e1-aafe-7a9761287be7/image.png)
- 서킷 브레이커를 서비스간 사이에 넣고 관련 함수들 래핑
- 네트워크 장애를 모니터링하여 특정 서비스가 timeou으로 설정된 시간 초과할 경우
  - 몇번의 요청을 한 이후에도 계속 timeout 시간을 초과하면 장애로 인식 (에러 반환)
- 서킷 브레이커가 trip 되며 그 이후 추가적인 호출은 발생X (fail fast)
  - trip : 서킷 브레이커 open 된 상태
  - closed 상태라면 기본적인 요청 수행하고, open된 상태라면 요청 수행하지 않고 빠르게 오류 반환

- 3) 서킷 브레이커 상태
  - closed[정상] : 네트워크 요청의 실패율이 임계치보다 낮음
  - open[오류] : 임계치 이상의 상태, 요청을 서비스로 전송하지 않고 바로 오류 반환 (fail fast)
  - half_open[확인중] : 장애가 해결되었는 지 확인을 위해 요청을 전송하여 응답 확인 (장애 풀리는지 확인 후 성공시 closed, 실패시 open)
  ![](https://velog.velcdn.com/images/24tngus/post/c1b7b279-dfe8-4844-a769-d58538edc0a7/image.png)

- 4) 구현
  - 넷플릭스의 [Hystrix](https://github.com/Netflix/Hystrix)와 [Resilience4j](https://github.com/resilience4j/resilience4j)

#### 6. 컨텐츠 관리

- 1) 불필요한 컨텐츠 제거
  - 불필요한 쿼리 등 제거
    - 인덱스를 추가하거나 쿼리를 변경해서 인덱스 사용하도록 개선
    - 개선이 안되는 쿼리라면 차라리 기능 OFF
    - `select *`로 조회하던 쿼리에서 실제 기능이 사용되지 않는 칼럼 모두 제거
  - 🗒️ 예시 : 인프런의 장애 복구 사례
    - 고용량의 컬럼을 강의 관련 쿼리에서 조회 항목에 포함시켜 가져옴
  ![](https://velog.velcdn.com/images/24tngus/post/18fe1e3b-7f98-4254-8d3e-e0f35c9cf4da/image.png)

- 2) CDN을 통한 컨텐츠 제공
  - 분산된 대규모 서버 네트워크를 기반으로 컨텐츠를 제공해서 서버 부하 줄임
  - CDN (Content Delivery Network) : 분산된 서버를 연결한 네트워크 
  
- 3) 컨텐츠 캐싱
  - 브라우저 캐시(쿠키, 로컬저장소, 세션저장소)를 통해 해당 요청에 관한 항목을 캐시에서 응답을 읽어 네트워크 요청에 관한 비용을 모두 제거
  
- 4) 컨텐츠 압축
  - 텍스트 기반 리소스는 gzip 또는 Brotli 통해 압축함 (70% 까지 압축)
  - 압축을 풀기 위해 서버에서 자원(CPU)를 사용하는 양 고려 필요
  
- 5) 컨텐츠 미리 준비된 응답
  - 시스템의 과도한 부하를 줄이기 위해 제공하는 컨텐츠 및 기능일 일시적으로 줄임
  - 방식 
    - 정적 테스트 페이지 제공 
    - 검색 비활성화
    - 더 적은 수의 검색 결과 반환
    - 필수적인지 않은 기능 비활성화

**Comment by jindream6128** [comment]
확인했습니다 !

**Comment by berryberries** [comment]
확인했습니다!
